src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs

```diff
@@ -1242,5 +1244,51 @@ public void Synchronized()
+        public void UseRegexConcurrently_ThreadSafe_Success(RegexOptions options, TimeSpan timeout)
+        {
+            const string Input = "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. Nunc viverra imperdiet enim. Fusce est. Vivamus a tellus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Proin pharetra nonummy pede. Mauris et orci. Aenean nec lorem. In porttitor. abcdefghijklmnx Donec laoreet nonummy augue. Suspendisse dui purus, scelerisque at, vulputate vitae, pretium mattis, nunc. Mauris eget neque at sem venenatis eleifend. Ut nonummy. Fusce aliquet pede non pede. Suspendisse dapibus lorem pellentesque magna. Integer nulla. Donec blandit feugiat ligula. Donec hendrerit, felis et imperdiet euismod, purus ipsum pretium metus, in lacinia nulla nisl eget sapien. Donec ut est in lectus consequat consequat. Etiam eget dui. Aliquam erat volutpat. Sed at lorem in nunc porta tristique. Proin nec augue. Quisque aliquam tempor magna. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Nunc ac magna. Maecenas odio dolor, vulputate vel, auctor ac, accumsan id, felis. Pellentesque cursus sagittis felis. Pellentesque porttitor, velit lacinia egestas auctor, diam eros tempus arcu, nec vulputate augue magna vel risus.nmlkjihgfedcbax";
+            const int Trials = 100;
+            const int IterationsPerTask = 10;
+
+            using var b = new Barrier(Environment.ProcessorCount);
+
+            for (int trial = 0; trial < Trials; trial++)
+            {
+                var r = new Regex("[a-q][^u-z]{13}x", options, timeout);
+                Task.WaitAll(Enumerable.Range(0, b.ParticipantCount).Select(_ => Task.Factory.StartNew(() =>
+                             {
+                                 b.SignalAndWait();
+                                 for (int i = 0; i < IterationsPerTask; i++)
+                                 {
+                                     Match m = r.Match(Input);
+                                     Assert.NotNull(m);
+                                     Assert.True(m.Success);
+                                     Assert.Equal("abcdefghijklmnx", m.Value);
+
+                                     m = m.NextMatch();
+                                     Assert.NotNull(m);
+                                     Assert.True(m.Success);
+                                     Assert.Equal("nmlkjihgfedcbax", m.Value);
+
+                                     m = m.NextMatch();
+                                     Assert.NotNull(m);
+                                     Assert.False(m.Success);
+                                 }
+                             }, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)).ToArray());
```

Comment:
I note Olli's hint that starting the regexes at the same moment is most likely to stress the DFA engine. But I wonder whether there might also be value in a variation of this that just plowed through the same 1000's of regex executions you're doing here, but without attempting to start them together. You'd get every possible variation of overlap between them.

Reply:
I'm not clear on what you're suggesting. Each thread here is doing 20 matches on the same Regex.  You're suggesting bumping that up to 1000?

Reply:
eh, it didn't make sense.

Reply:
Hah, ok :)